## Author: Saloni Gupta 
<br/>
Date: 22/10/2020 <br/>
Heroes Of Pymoli Data Analysis

In [25]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load, encoding='utf-8')
#with pd.option_context('display.colheader_justify','left'):
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [2]:
# calculate total players count, by extracting the unique player list
# by using .unique() and counting them by using len()
Total_Players = len(purchase_data.SN.unique())
# create a data frame with Total_Players count
Total_players =pd.DataFrame({"Total Players": Total_Players},index=[0])
Total_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Purchasing Analysis (Total) #
# calculate the count of unique items 
unique_item = len(purchase_data["Item ID"].unique())
# calculate the average price of the purchase 
average_price = purchase_data["Price"].mean()
# caluclate the total number of purchases made
total_purchases = purchase_data["Purchase ID"].count()
# calculate the total revenue
total_price = purchase_data["Price"].sum()
# create a summary data frame with the above results and do a clean formatting 
Summary =pd.DataFrame({"Number of Unique Items": unique_item,"Average Price":average_price,
                       "Total Purchases":total_purchases,"Total Revenue":total_price},index=[0])
Summary["Average Price"] = Summary["Average Price"].map("${:.2f}".format)
Summary["Total Revenue"] = Summary["Total Revenue"].map("${:,.2f}".format)
Summary

Number of Unique Items Average Price  Total Purchases Total Revenue
0                     179         $3.05              780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
# Gender Demographics

# Percentage and Count of Male Players
# Percentage and Count of Female Players
# Percentage and Count of Other / Non-Disclosed
gender_grouped = purchase_data.groupby("Gender")
unique_members = gender_grouped["SN"].nunique()
percentage = (unique_members/Total_Players)*100

# Create a data frame with the summary gender data
gender_summary = pd.DataFrame({"Total Counts": unique_members,"Percentage of Players": percentage})

gender_summary["Percentage of Players"] = gender_summary["Percentage of Players"].map("{:.2f}%".format)

gender_summary = gender_summary.sort_values(["Total Counts"],ascending = False)
gender_summary.index.name = None
gender_summary

Total Counts Percentage of Players
Male                            484                84.03%
Female                           81                14.06%
Other / Non-Disclosed            11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
# Purchasing Analysis(Gender)
# Do a groupby on gender and calculated purchase count, avg. purchase price, avg. purchase total per person 

purchase_count = gender_grouped["Purchase ID"].count()
avg_purchase = gender_grouped["Price"].mean()
total_purchase_value = gender_grouped["Price"].sum()
avg_purchase_person =(total_purchase_value/unique_members)

# create a summary DataFrame to hold the above results 
purchasing_summary_df = pd.DataFrame({"Purchase count":purchase_count,"Average Purchase Price":avg_purchase,
                           "Total Purchase value":total_purchase_value,"Average Purchase Total per person":avg_purchase_person})
# do a clean formatting 
purchasing_summary_df["Average Purchase Price"] = purchasing_summary_df["Average Purchase Price"].map("${:.2f}".format)
purchasing_summary_df["Total Purchase value"]  = purchasing_summary_df["Total Purchase value"].map("${:,.2f}".format)
purchasing_summary_df["Average Purchase Total per person"] = purchasing_summary_df["Average Purchase Total per person"]
                                                            .map("${:.2f}".format)
purchasing_summary_df

Purchase count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase value Average Purchase Total per person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
# Age Demographics
# Establish bins for ages and create labels
age_bins = [0, 9, 14, 19, 24, 29, 34, 39, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# slice the data using pd.cut and Categorize the existing players based on age_bins  
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"],age_bins, labels=group_names)
purchase_data
# do a groupby on Age Ranges
age_grouped = purchase_data.groupby("Age Ranges")
#calculate total count and percentage of players by age category
unique_members = age_grouped["SN"].nunique()
unique_members
Percent_of_players = (unique_members/Total_Players)*100
Percent_of_players
# create a DataFrame to hold the above results 
age_demographics = pd.DataFrame({"Total count":unique_members,"Percentage of players":Percent_of_players})
# change the index to none and do clean formatting 
age_demographics.index.name = None
age_demographics["Percentage of players"] = age_demographics["Percentage of players"].map("{:.2f}%".format)
age_demographics

Total count Percentage of players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
# Purchasing Analysis(Age)

# do calculations to obtain purchase count, avg. purchase price, total purchase value , average purchase per person
# do a groupby on Age Group 
age_grouped = purchase_data.groupby("Age Ranges")
purchase_count = age_grouped["Age Ranges"].count()
avg_purchase = age_grouped["Price"].mean()
total_purchase_value = age_grouped["Price"].sum()
avg_purchase_person =(total_purchase_value/unique_members)

# create a DataFrame to hold the above results 
Purchase_analysis_age = pd.DataFrame({"Purchase count":purchase_count,"Average Purchase Price":avg_purchase,
                           "Total Purchase Value":total_purchase_value,"Average Purchase Total per person":avg_purchase_person})
# do a clean formatting 
Purchase_analysis_age["Average Purchase Price"] = Purchase_analysis_age["Average Purchase Price"].map("${:.2f}".format)
Purchase_analysis_age["Total Purchase Value"]  = Purchase_analysis_age["Total Purchase Value"].map("${:,.2f}".format)
Purchase_analysis_age["Average Purchase Total per person"] = Purchase_analysis_age["Average Purchase Total per person"]
                                                            .map("${:.2f}".format)
Purchase_analysis_age

Purchase count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Average Purchase Total per person  
Age Ranges                                    
<10                                    $4.54  
10-14                                  $3.76  
15-19                                  $3.86  
20-24                                  $4.32  
25-29                                  $3.81  
30-34                                  $4.12  
35-39                                  $4.76  
40+                                    $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
# Top Spenders 
# do a groupby on (SN-Screen Name) and assign it to a variable 
SN_grouped = purchase_data.groupby("SN")

# do calculations to obtain purchase count, avg. purchase price, total purchase value 
purchase_count = SN_grouped["Price"].count()
avg_purchase = SN_grouped["Price"].mean()
total_purchase_value = SN_grouped["Price"].sum()

# create a DataFrame to hold the results 
Top_spenders = pd.DataFrame({"Purchase count":purchase_count,"Average Purchase Price":avg_purchase,
                           "Total Purchase value":total_purchase_value})

# Sort the total purchase value column in descending order 
Top_spenders = Top_spenders.sort_values("Total Purchase value", ascending = False)

# do a clean formatting 
Top_spenders["Average Purchase Price"] = Top_spenders["Average Purchase Price"].map("${:.2f}".format)
Top_spenders["Total Purchase value"]  = Top_spenders["Total Purchase value"].map("${:,.2f}".format)

# display the above results(Top Spenders sumamry table )
Top_spenders.head()

Purchase count Average Purchase Price Total Purchase value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
#Most Popular Items
# Retrieve the Item ID, Item Name, and Item Price columns (using .loc)  
popular_df = purchase_data.loc[:,["Item ID","Item Name","Price"]]
#dfStyler = popular_df.style.set_properties(**{'text-align': 'left'})
#dfStyler.set_table_styles([dict(selector='th', props=[('text-align', 'left')])])
#DataFrame.align(left)

# Group by Item ID and Item Name. 
item_grouped = purchase_data.groupby(["Item ID","Item Name"])

# Perform calculations to obtain purchase count, item price, and total purchase value
purchase_count = item_grouped["Price"].count()
avg_purchase = item_grouped["Price"].mean()
total_purchase_value = item_grouped["Price"].sum()

# create a DataFrame to hold the results 
most_popular_items = pd.DataFrame({"Purchase Count":purchase_count,"Average Purchase Price":avg_purchase,
                                   "Total Purchase Value":total_purchase_value})

# Sort the purchase count  column in descending order 
most_popular_items = most_popular_items.sort_values("Purchase Count", ascending = False)

# do a clean formatting 
most_popular_items["Total Purchase Value"]  = most_popular_items["Total Purchase Value"].map("${:,.2f}".format)
most_popular_items["Average Purchase Price"] = most_popular_items["Average Purchase Price"].map("${:.2f}".format)


# display the Most Popular Items summary table 
most_popular_items.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
92      Final Critic                                                  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
145     Fiery Glass Crusader                                          $4.58   
132     Persuasion                                                    $3.22   
108     Extraction, Quickblade Of Trembling Hands                     $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [24]:
#Most Popular Items
# Retrieve the Item ID, Item Name, and Item Price columns (using .loc)  
popular_df = purchase_data.loc[:,["Item ID","Item Name","Price"]]

# Group by Item ID and Item Name. 
item_grouped = purchase_data.groupby(["Item ID","Item Name"])

# Perform calculations to obtain purchase count, item price, and total purchase value
purchase_count = item_grouped["Price"].count()
avg_purchase = item_grouped["Price"].mean()
total_purchase_value = item_grouped["Price"].sum()

# create a DataFrame to hold the results 
most_popular_items = pd.DataFrame({"Purchase Count":purchase_count,"Average Purchase Price":avg_purchase,
                                   "Total Purchase Value":total_purchase_value})

# Sort the purchase count  column in descending order 
most_popular_items = most_popular_items.sort_values("Total Purchase Value", ascending = False)

# do a clean formatting 
most_popular_items["Total Purchase Value"]  = most_popular_items["Total Purchase Value"].map("${:,.2f}".format)
most_popular_items["Average Purchase Price"] = most_popular_items["Average Purchase Price"].map("${:.2f}".format)

#dfStyler = most_popular_items.style.set_properties({'text-align': 'left'})
# display the Most Popular Items summary table 
most_popular_items.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Average Purchase Price  \
Item ID Item Name                                                             
92      Final Critic                                                  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm                  $4.23   
82      Nirvana                                                       $4.90   
145     Fiery Glass Crusader                                          $4.58   
103     Singed Scalpel                                                $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

### Conclusions:-
1) Of the 780 players list, there are 576 unique players, the vast majority are male (84%) and smaller proportion of female players (14%).<br/>
2) Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.6%) and 25-29 (13%).<br/>
3) The majority of purchases are also done by the age group 20-24 with total purchase value of USD 1114 although maximum purchase of 5 items is done by Lisosia93 who falls in age group 25-29. <br/>
4) Out of 183 items on purchase list, the item  Final Critic is the most profitable as well as the most popular item with total purchase value USD 59.99 having sold max 13 items, followed by Oathbreaker, Last Hope of the Breaking Storm with total purchase value USD 50.76 ,selling 12 items. <br/>
5) Since the participating players were predominantly male 484 compared to 81 females, the majority of purchases are done by Male with no significant difference in the total item purchased by male(652/484) vs total item purchased by female(113/81).